game.js

```diff
@@ -0,0 +1,148 @@
+let bot = require("./bot.js");
+let helpers = require("./helpers.js");
+let lastLetter = "";
+let cities = [
+  "Архангельск",
+  "Борисполь",
+  "Воронеж",
+  "Гюмри",
+  "Донецк",
+  "Енакиево",
+  "Жуковский",
+  "Запорожье",
+  "Изюм",
+  "Йошкар-Ола",
+  "Киев",
+  "Лондон",
+  "Москва",
+  "Николаев",
+  "Орёл",
+  "Париж",
+  "Ростов",
+  "Селидово",
+  "Трускавец",
+  "Уфа",
+  "Флоренция",
+  "Харьков",
+  "Цюрих",
+  "Челябинск",
+  "Шанхай",
+  "Щёлково",
+  "Ыгдыр",
+  "Электросталь",
+  "Юрмала",
+  "Ялта"
+].map(item => item.toLowerCase());
+
+exports.session = class Session {
+  constructor(id, chatId = 0) {
+    this.id = id;
+    this.chatId = chatId;
+    this.spentCities = new Set();
+  }
+};
+
+function randomCity(arrCities) {
+  return arrCities[helpers.getRandomInt(arrCities.length)];
+}
+
+function lastValidLetter(str) {
+  let lastLetter = str[str.length - 1];
+  let invalidLetters = ["ь", "Ъ"];
+  let i = 2;
+
+  while (invalidLetters.includes(lastLetter) || str.length == i) {
+    lastLetter = str[str.length - i];
+    i++;
+  }
+  return lastLetter;
+}
+
+function checkCityInDB(sessionId, city, letter) {
+  let result = { isError: false, errorMsg: "" };
+
+  if (letter != city[0]) {
+    result.errorMsg = "Нужно назвать город на букву " + letter.toUpperCase();
+    result.isError = true;
+    return result;
+  }
+  if (bot.session[sessionId].spentCities.has(city)) {
+    result.errorMsg = "Этот город уже был назван!";
+    result.isError = true;
+  }
+  return result;
+}
+
+function selectCityByLetter(sessionId, letter, cities) {
+  let findCities = [];
+  findCities = cities.filter(
+    item => item[0] == letter && !bot.session[sessionId].spentCities.has(item)
+  );
+  //Условие ниже не срабатывает.. когда фильтр ничего не находит он должен вернуть пустой массив,
```

Comment:
I don't understand this comment

Reply:
```suggestion
  function selectCityByLetter(sessionId, letter, cities) {
  let findCities = [];
  findCities = cities.filter(
    item => item[0] == letter && !bot.session[sessionId].spentCities.has(item)
  );
  //The condition below does not work .. when the filter does not find anything, it should return an empty array,
  //but checking for an empty array for some reason does not work
  if (findCities === [] || findCities === undefined || findCities === NaN) {
    //This conditions does not work..
    console.log("game.js 84 Город не найден"); //This console.log should be display but nothing happens, and this function returns undefined
    return -1;
  } else return randomCity(findCities);
}
```

Reply:
Ok, i fixed it. I add a conditions `findCities.length === 0` and it's worked... But strange that conditions `findCities === []` doesn't work. In the JS book tells that `arr.filter` return a blank array if nothing found.. Why it does not work, I don't understand..